osu.Game.Rulesets.Osu/Difficulty/Skills/Tap.cs

```diff
@@ -0,0 +1,161 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+using MathNet.Numerics;
+using MathNet.Numerics.LinearAlgebra;
+
+using osu.Game.Rulesets.Osu.Difficulty.MathUtil;
+using osu.Game.Rulesets.Osu.Objects;
+
+
+namespace osu.Game.Rulesets.Osu.Difficulty.Skills
+{
+    public static class Tap
+    {
+
+        private const int mash_level_count = 11;
+        private const double spaced_buff_factor = 0.10;
+
+        private static readonly Vector<double> decay_coeffs = Vector<double>.Build.Dense(Generate.LinearSpaced(4, 2.3, -2.8))
+                                                                                 .PointwiseExp();
+
+        private static readonly double[] timescale_factors = { 1.02, 1.02, 1.05, 1.15 };
+
+        public static (double, double, double[], double[], List<Vector<double>>) CalculateTapAttributes
+            (List<OsuHitObject> hitObjects, double clockRate)
+        {
+            (var strainHistory, var tapDiff) = calculateTapStrain(hitObjects, 0, clockRate);
+            double burstStrain = strainHistory.Max(v => v[0]);
+
+            var streamnessMask = CalculateStreamnessMask(hitObjects, burstStrain, clockRate);
+            double streamNoteCount = streamnessMask.Sum();
+
+            (var mashLevels, var tapSkills) = calculateMashLevelsVSTapSkills(hitObjects, clockRate);
+
+            return (tapDiff, streamNoteCount, mashLevels, tapSkills, strainHistory);
+        }
+
+        /// <summary>
+        /// Calculates the strain values at each note and the maximum strain values
+        /// </summary>
+        private static (List<Vector<double>>, double) calculateTapStrain(List<OsuHitObject> hitObjects,
+                                                                                 double mashLevel,
+                                                                                 double clockRate)
+        {
+            var strainHistory = new List<Vector<double>> { decay_coeffs * 0, decay_coeffs * 0 };
+            var currStrain = decay_coeffs * 1;
```

Comment:
isn't this just effectively `decay_coeffs`?

Reply:
it should be `decay_coeffs * 0` because it is the initial strain value